File: zh-Hant/codes/cpp/chapter_hashing/built_in_hash.cpp
/**
 * File: built_in_hash.cpp
 * Created Time: 2023-06-21
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Driver Code */
int main() {
    int num = 3;
    size_t hashNum = hash<int>()(num);
    cout << "整數 " << num << " 的雜湊值為 " << hashNum << "\n";

    bool bol = true;
    size_t hashBol = hash<bool>()(bol);
    cout << "布林量 " << bol << " 的雜湊值為 " << hashBol << "\n";

    double dec = 3.14159;
    size_t hashDec = hash<double>()(dec);
    cout << "小數 " << dec << " 的雜湊值為 " << hashDec << "\n";

    string str = "Hello 演算法";
    size_t hashStr = hash<string>()(str);
    cout << "字串 " << str << " 的雜湊值為 " << hashStr << "\n";

    // 在 C++ 中，內建 std:hash() 僅提供基本資料型別的雜湊值計算
    // 陣列、物件的雜湊值計算需要自行實現
}

File: zh-hant/codes/ruby/chapter_hashing/built_in_hash.rb
=begin
File: built_in_hash.rb
Created Time: 2024-04-13
Author: Xuan Khoa Tu Nguyen (ngxktuzkai2000@gmail.com)
=end

require_relative '../utils/list_node'

### Driver Code ###
if __FILE__ == $0
  num = 3
  hash_num = num.hash
  puts "整數 #{num} 的雜湊值為 #{hash_num}"

  bol = true
  hash_bol = bol.hash
  puts "布林量 #{bol} 的雜湊值為 #{hash_bol}"

  dec = 3.14159
  hash_dec = dec.hash
  puts "小數 #{dec} 的雜湊值為 #{hash_dec}"

  str = "Hello 演算法"
  hash_str = str.hash
  puts "字串 #{str} 的雜湊值為 #{hash_str}"

  tup = [12836, '小哈']
  hash_tup = tup.hash
  puts "元組 #{tup} 的雜湊值為 #{hash_tup}"

  obj = ListNode.new(0)
  hash_obj = obj.hash
  puts "節點物件 #{obj} 的雜湊值為 #{hash_obj}"
end
